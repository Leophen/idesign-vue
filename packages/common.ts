import _ from 'lodash'
import {
  Component,
  getCurrentInstance,
  Slots,
  VNode,
} from 'vue'

// 通用位置类型
export type placementType =
  | 'top'
  | 'top-left'
  | 'top-right'
  | 'bottom'
  | 'bottom-left'
  | 'bottom-right'
  | 'left'
  | 'left-top'
  | 'left-bottom'
  | 'right'
  | 'right-top'
  | 'right-bottom'

// 通用触发类型
export type triggerType = 'hover' | 'click' | 'context-menu'

// 转换 width/height/left/... 样式值
export const turnValue = (val: string | number | undefined) => {
  if (!val) {
    return
  }
  if (_.isNumber(val as number)) {
    return `${val}px`
  }
  return val
}

// const _stateValue = ref(defaultValue)
// const innerValue = computed(() => value ?? _stateValue.value)

/**
 * 渲染 default slot，获取子组件 VNode。处理多种子组件创建场景
 * @returns {function(childComponentName: string, slots: Slots): VNode[]}
 * @param childComponentName
 * @param slots
 * @example const getChildByName = useChildComponentSlots()
 * @example getChildComponentByName('xxItem')
 */
export const useChildComponentSlots = () => {
  const instance = getCurrentInstance()
  return (childComponentName: string, slots?: Slots): VNode[] => {
    if (!slots) {
      slots = instance?.slots
    }
    const content = slots?.default?.() || []

    return (
      content
        .map((item: VNode) => {
          if (item.children && Array.isArray(item.children))
            return item.children
          return item
        })
        .flat()
        // @ts-ignore
        .filter((item: VNode) =>
          (item.type as Component).name?.endsWith(childComponentName)
        ) as VNode[]
    )
  }
}
