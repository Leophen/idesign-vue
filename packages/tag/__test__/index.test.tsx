import { vi } from 'vitest';
import { Tag } from '../index';
import { render } from '../../common'

describe('Tag 组件测试', () => {
  it('create', () => {
    const wrapper = render(<Tag>标签</Tag>);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('type', () => {
    const wrapper = render(<Tag type="success">标签</Tag>);
    expect(wrapper.classes()).toContain('i-tag--type-success');
  });

  it('theme', () => {
    const wrapper = render(<Tag theme="dark">标签</Tag>);
    expect(wrapper.classes()).toContain('i-tag--theme-dark');
  });

  it('icon', () => {
    const wrapper = render(<Tag icon="TagPrice">标签</Tag>);
    expect(wrapper.find('.icon-TagPrice').exists()).toBe(true);
  });

  it('onAdd', () => {
    const clickFn = vi.fn();
    const wrapper = render(<Tag addable onAdd={clickFn}>标签</Tag>);
    expect(wrapper.find('.i-icon').classes()).toContain('icon-ThePlus');
    wrapper.find('.i-tag').trigger('click');
    expect(clickFn).toHaveBeenCalled();
  });

  it('onClose', () => {
    const clickFn = vi.fn();
    const wrapper = render(<Tag closeable onClose={clickFn}>标签</Tag>);
    expect(wrapper.find('.i-tag--close-btn').classes()).toContain('i-tag--close-btn');
    wrapper.find('.i-tag--close-btn').trigger('click');
    expect(clickFn).toHaveBeenCalled();
  });

  it('size', () => {
    const wrapper = render(<Tag size="small">标签</Tag>);
    expect(wrapper.classes()).toContain('i-tag--size-small');
  });

  it('maxWidth', () => {
    const wrapper = render(<Tag maxWidth={100}>标签</Tag>);
    expect(wrapper.element.getAttribute('style')).toMatch('max-width: 100px;');
  });
});

