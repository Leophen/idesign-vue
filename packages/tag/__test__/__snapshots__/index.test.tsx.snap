// Vitest Snapshot v1

exports[`Tag 组件测试 > create 1`] = `
<div
  class="i-tag i-tag--type-default"
>
  <!--v-if-->
  <!--v-if-->
  
  标签
  
  <!--v-if-->
</div>
`;
